sympy/simplify/simplify.py

```diff
@@ -1101,6 +1106,56 @@ def bottom_up(rv, F, atoms=False, nonbasic=False):
     return rv
 
 
+def kroneckersimp(expr):
+    """
+    Simplify expressions with KroneckerDelta.
+
+    The only simplification currently attempted is to identify multiplicative cancellation:
+
+    >>> from sympy import KroneckerDelta, kroneckersimp
+    >>> from sympy.abc import i, j
+    >>> kroneckersimp(1 + KroneckerDelta(0, j) * KroneckerDelta(1, j))
+    1
+    """
+    def args_cancel(args1, args2):
+        for i1 in range(2):
+            for i2 in range(2):
+                a1 = args1[i1]
+                a2 = args2[i2]
+                a3 = args1[(i1 + 1) % 2]
+                a4 = args2[(i2 + 1) % 2]
+                if Eq(a1, a2) is S.true and Eq(a3, a4) is S.false:
+                    return True
+        return False
+
+    def cancel_kronecker_mul(m):
+        from sympy.utilities.iterables import subsets
+
+        args = m.args
+        deltas = [a for a in args if isinstance(a, KroneckerDelta)]
+        for delta1, delta2 in subsets(deltas, 2):
+            args1 = delta1.args
+            args2 = delta2.args
+            if args_cancel(args1, args2):
+                return 0*m
+        return m
+
+    if not expr.has(KroneckerDelta):
+        return expr
+
+    if expr.has(Piecewise):
+        expr = expr.rewrite(KroneckerDelta)
```

Comment:
This is the only thing I'm not sure about here. Can there be some situations where the Piecewise form is preferred to KroneckerDelta? It looks like in some situations this will rewrite Piecewises to KroneckerDelta but not do any simplification after and I'm not sure if it is always good for simplify to do that.

I though that this would be an example but it wasn't:
```julia
In [3]: expr = KroneckerDelta(n, 0) + Piecewise((y, Eq(x, 0)), (x, True))                                                         

In [4]: expr                                                                                                                      
Out[4]: 
⎛⎧y  for x = 0⎞       
⎜⎨            ⎟ + δ   
⎝⎩x  otherwise⎠    0,n

In [5]: simplify(expr)                                                                                                            
Out[5]: 
⎛⎧y  for x = 0⎞       
⎜⎨            ⎟ + δ   
⎝⎩x  otherwise⎠    0,n

In [7]: kroneckersimp(expr)                                                                                                       
Out[7]: 
x⋅⎛1 - δ   ⎞ + y⋅δ    + δ   
  ⎝     0,x⎠      0,x    0,n
```
Does that simplification get rejected higher up in simplify?

If so then I guess this is fine.

Reply:
Yes, this can happen: I added a test that shows this already (simplify(K(1, i) + Piecewise((1, Eq(j, 2)), (0, True)))).